packages/core/src/stateUtils.ts

```diff
@@ -1743,53 +1485,43 @@ export function resolveActionsAndContext<
           {
             machine,
             state: currentState!,
-            action: actionObject
+            action: actionObject,
+            actorContext: actorCtx
           }
         );
 
         if (
           resolvedActionObject.type === actionTypes.raise ||
           (resolvedActionObject.type === actionTypes.send &&
-            resolvedActionObject.params.to === SpecialTargets.Internal)
+            actorCtx &&
+            (resolvedActionObject as SendActionObject).params.to ===
+              actorCtx.self)
         ) {
           raiseActions.push(resolvedActionObject);
         } else {
           resolvedActions.push(resolvedActionObject);
-          if (
-            predictableExec &&
-            resolvedActionObject.type !== actionTypes.invoke
-          ) {
-            predictableExec(
-              resolvedActionObject,
-              preservedContexts[preservedContexts.length - 1],
-              _event
-            );
+          // TODO: only using actorCtx.exec as a flag to execute; actually use it for execution
+          if (actorCtx) {
+            execAction(resolvedActionObject, currentState, actorCtx);
           }
         }
       }
       return;
     }
-    const contextIndex = preservedContexts.length - 1;
     if (isExecutableAction(executableActionObject)) {
-      executableActionObject.setContext(preservedContexts[contextIndex]);
-      resolvedActions.push(executableActionObject);
-    } else {
-      const resolvedActionObject = toActionObject(
-        executableActionObject,
-        machine.options.actions
-      );
-
-      const actionExec = new ExecutableAction(resolvedActionObject);
-      actionExec.setContext(preservedContexts[contextIndex]);
-
-      resolvedActions.push(actionExec);
+      executableActionObject.setContext(updatedContext);
     }
-
-    if (predictableExec) {
-      predictableExec(
-        resolvedActions[resolvedActions.length - 1],
-        preservedContexts[preservedContexts.length - 1],
-        _event
+    resolvedActions.push(executableActionObject);
+    const resolvedAction = resolvedActions[resolvedActions.length - 1];
+    // TODO: only using actorCtx.exec as a flag to execute; actually use it for execution
+    if (actorCtx) {
+      execAction(
+        resolvedAction,
+        currentState.clone({
+          context: updatedContext,
+          _event
```

Comment:
do we need to assign the new `_event` here too? I would assume that it's "stable" for all the actions executed here

I also wonder why do we only `.clone` here. Shouldn't this be handled in the same way for all action types?